src/main/scala/chisel3/util/MemInterface.scala

```diff
@@ -0,0 +1,254 @@
+package chisel3.util
+
+import chisel3._
+
+import chisel3.internal.Builder
+import chisel3.experimental.SourceInfo
+import chisel3.internal.sourceinfo.{MemTransform, SourceInfoTransform}
+import scala.language.reflectiveCalls
+import scala.language.experimental.macros
+
+/** A bundle of signals representing a memory read port.
+  *
+  * @tparam tpe The data type of the memory port
+  * @param addrWidth The width of the address signal
+  */
+class MemoryReadPort[T <: Data](tpe: T, addrWidth: Int) extends Bundle {
+  val addr = Input(UInt(addrWidth.W))
+  val enable = Input(Bool())
+  val data = Output(tpe)
+}
+
+/** A bundle of signals representing a memory write port.
+  *
+  * @tparam tpe The data type of the memory port
+  * @param addrWidth The width of the address signal
+  * @param masked Whether this read/write port should have an optional mask.
+  *
+  * @note `masked` is only valid if tpe is a Vec; if this is not the case no mask will be initialized
+  *       regardless of the value of `masked`.
+  */
+class MemoryWritePort[T <: Data] private[chisel3] (tpe: T, addrWidth: Int, masked: Boolean) extends Bundle {
+  val addr = Input(UInt(addrWidth.W))
+  val enable = Input(Bool())
+  val data = Input(tpe)
+  val mask: Option[Vec[Bool]] = if (masked) {
```

Comment:
IMO, chisel type might be too high level for memories. Even for `tpe: T`(I personally don't like it since it cannot be serialized).
For all memories I have encountered in ASIC designs, they are all bit level, so I think dataWidth mask is enough for it.

Reply:
Regular Chisel type is fine, if you don't want to use it, you can just use `UInt` or `Vec[UInt]` but theres no reason to define it away. Per the memories being bit-level, we can of course just pack the data type into bitvectors in Verilog.